embassy-stm32/src/aes/mod.rs

```diff
@@ -0,0 +1,743 @@
+        if ccf {
+            T::regs().cr().modify(|w| w.set_ccfie(false));
+            AES_WAKER.wake();
+        }
+    }
+}
+
+/// AES-CCM Cipher Mode
+pub struct AesCcm<
+    'c,
+    'd,
+    const KEY_SIZE: usize,
+    const TAG_SIZE: usize,
+    const IV_SIZE: usize,
+    T: Instance,
+    DmaIn: 'static,
+    DmaOut: 'static,
+> {
+    aes: &'c mut Aes<'d, T, DmaIn, DmaOut>,
+    key: &'c [u8; KEY_SIZE],
+    aad_header_buffer: [u8; 8],
+    aad_header_len: usize,
+    payload_len: usize,
+    iv: [u8; 16],
+    dir: Direction,
+    aad_processed: bool,
+}
+
+impl<'c, 'd, const KEY_SIZE: usize, const TAG_SIZE: usize, const IV_SIZE: usize, T: Instance, DmaIn, DmaOut>
+    AesCcm<'c, 'd, KEY_SIZE, TAG_SIZE, IV_SIZE, T, DmaIn, DmaOut>
+{
+    /// Constructs a new AES-CCM cipher for a cryptographic operation.
+    pub fn new(
+        aes: &'c mut Aes<'d, T, DmaIn, DmaOut>,
+        key: &'c [u8; KEY_SIZE],
+        nonce: &'c [u8; IV_SIZE],
+        aad_len: usize,
+        payload_len: usize,
+        dir: Direction,
+    ) -> Self {
+        let mut aad_header_buffer: [u8; 8] = [0; 8];
+        let aad_header_len = Self::setup_aad_header(&mut aad_header_buffer, aad_len);
+        let mut iv: [u8; 16] = [0; 16];
+        Self::setup_iv(&mut iv, nonce, aad_len, payload_len);
+
+        return Self {
+            aes,
+            key,
+            aad_header_buffer,
+            aad_header_len,
+            iv,
+            payload_len,
+            dir,
+            aad_processed: false,
+        };
+    }
+
+    /// Prepares header which will be sent at the beginning of first AAD block.
+    ///
+    /// Returns length of the header.
+    fn setup_aad_header(aad_header_buffer: &mut [u8; 8], aad_len: usize) -> usize {
+        if aad_len == 0 {
+            return 0;
+        }
+
+        if aad_len < 65280 {
+            aad_header_buffer[0] = (aad_len >> 8) as u8 & 0xFF;
+            aad_header_buffer[1] = aad_len as u8 & 0xFF;
+            return 2;
+        } else {
+            aad_header_buffer[0] = 0xFF;
+            aad_header_buffer[1] = 0xFE;
+            let aad_len_bytes: [u8; 4] = aad_len.to_be_bytes();
+            aad_header_buffer[2] = aad_len_bytes[0];
+            aad_header_buffer[3] = aad_len_bytes[1];
+            aad_header_buffer[4] = aad_len_bytes[2];
+            aad_header_buffer[5] = aad_len_bytes[3];
+            return 6;
+        }
+    }
+    /// Prepares payload that will land in IV register.
+    /// For AES CCM, IV register is filled with payload called **block 0**.
+    /// For details, refer to CCM specification in RFC 3610.
+    pub fn setup_iv(block0: &mut [u8; 16], iv: &'c [u8; IV_SIZE], aad_len: usize, payload_len: usize) {
+        if aad_len > 0 {
+            block0[0] = 0x40;
+        }
+        block0[0] |= ((((TAG_SIZE as u8) - 2) >> 1) & 0x07) << 3;
+        block0[0] |= ((15 - (iv.len() as u8)) - 1) & 0x07;
+
+        block0[1..1 + iv.len()].copy_from_slice(iv);
```

Comment:
Maybe pedantic, but this is a panic path. Is there any merit in using a `static_assertions::const_assert` to guard against the panic?

e.g.
```rust
const_assert!(IV_SIZE < 16);
```

That would at least fail compilation if we are being dumb.